pyquil/quilbase.py

```diff
@@ -199,11 +199,23 @@ def out(self) -> str:
                 _format_qubits_out(self.qubits),
             )
 
-    def controlled(self, control_qubit: QubitDesignator) -> "Gate":
+    def controlled(self, control_qubits: Union[QubitDesignator, List[QubitDesignator]]) -> "Gate":
```

Comment:
Changing the parameter name here is unfortunately a backwards incompatible change.

Neither of the options I can think of for maintaining backwards compat are especially appealing. Either we leave the parameter name alone, or else provide a new keyword-only parameter with the old name and make the positional parameter optional, and raise a deprecation warning if callers supply the argument under the old keyword name and an error if callers supply neither.

As convoluted as it sounds, the second option seems preferable since it at least gives us a path to deprecate and eventually remove the old keyword name.

Reply:
I don't have much of a problem with `control_qubit` possibly being plural. The name doesn't suggest you can give more than one control, but the type hint does and I would expect most devtools to surface that information.

Reply:
Agreed. Maybe we should just leave a comment explaining that the inconsistent naming is to preserve backwards compat.

Reply:
I have not tested this but it looks like the definition could be simplified as:
```python
def controlled(self, *control_qubits: Sequence[QubitDesignator]) -> "Gate":
    """
    Add the CONTROLLED modifier to the gate with the given control qubit or list of control
    qubits.
    """
    if len(control_qubits) == 0:
        return self

    self.modifiers.insert(0, "CONTROLLED")
    self.qubits.insert(0, unpack_qubit(control_qubits[0]))
    
    return self.controlled(control_qubits[1:])
```

Perhaps the type should be `Iterable` or `Sized` rather than `Sequence`, though.

Reply:
I hit the button too quickly!! Checking for the length of the whole list at each step of the iteration is not a good idea (although unlikely to have a big impact here). It is better to declare `control_qubits` as `Iterable` and use `next` and then catch `StopIteration` or something to that effect in order to be O(n).

Reply:
:)

I almost made a similar comment recommending instead something like:

```python
new_qubits = [unpack_qubit(q) for q in control_qubits]
new_modifers = ["CONTROLLED"] * len(control_qubits)
self.modifers = new_modifers + self.modifers
self.qubits = new_qubits + self.qubits
```

in the end I deleted it because efficiency (as you said) is unlikely to be practically important here, since `len(control_qubits)` is probably small--otherwise you will have bigger problems--and I wasn't 100% sure nothing depends on `self.modifiers` and `self.qubits` being mutated in place, and I was too lazy to go looking. It seems pretty unlikely anything depends on the mutation and if it does it shouldn't, but on the balance the performance penalty seemed like a small price to pay to keep the behavior exactly the same as before.

Reply:
To be clear, I wouldn't be opposed to switching to something O(n) here or even changing the behavior w.r.t. mutation if others feel like it's worthwhile.

Reply:
@appleby I concur. I still think that adding the asterisk to the `control_qubits` parameter leads to slightly clearer code (no need for `isinstance`).

Reply:
I agree the asterisk improves the interface for `controlled`. The downside is that it would be awkward to present the same interface for `forked` (assuming we want support something similar for `forked`), since `forked` would then expect both a sequence of control qubits and a list (or should it be a list of lists?) of forked parameters.

Reply:
I will fix that tomorrow, thanks! I can fix that for forked too, but to be honest I didn't get the idea of forked. How it modifies the gate? I couldn't understand it from docs.

Reply:
If you think of the `controlled` modifier as conditionally applying a gate (or not) based on the state of the control qubit, then you can think of `forked` as conditionally applying a gate with one of two sets of parameters where the set of parameters chosen depends on the state of the control qubit.

For example,

```python
Program(RX(np.pi/2, 1).forked(0, [np.pi]))
```

is equivalent to `RX(np.pi/2, 1)` if qubit 0 is in the `|0>` state, and to `RX(np.pi, 1)` if qubit 0 is in the `|1>` state (and some combination of the two if the control qubit is in an arbitrary superposition state).

Reply:
Oh, I see. But why in forked You don't have a gate passed? For example
```
Program(RX(np.pi/2, 1).forked(0, X)
```
or, using Your example
```
Program(RX(np.pi/2, 1).forked(0,  lambda x: RX(np.pi, x))
```
or another version that would require assertion on target qubit match

```
Program(RX(np.pi/2, 1).forked(0,  RX(np.pi, 1))
```
Each version is way more general, and yet still is quite simple. 

Sorry for a lot of questions, but after that, I will know what to do with forked modifier:)

Reply:
I guess I see now: it makes it difficult when stacking `forked` modifier.
Let me then fixed `controlled` only.

Reply:
`*control_qubits: Sequence[QubitDesignator]` means a tuple of Sequence. To my best knowledge, it is impossible to have at one time `X(0).controlled(1, 2)` and `X(0).controlled([1,2])` without assertion or allowing `X(0).controlled([1,2],[3,4])`. What would be an expected behavior? I personally vote for `*control_qubits: QubitDesignator` and not allow passing any iterable directly.

Reply:
My vote is to only support 2 (and of course the pre-existing convention of passing a single number, i.e. `X(0).controlled(1)`).

You raise a good point, which is that (4) is not really useful since a set is unordered. Hence my comment elsewhere about being able to broaden the input type from `Sequence` to `Iterable` was wrong. We really do want an ordered `Sequence` here.

My reason for preferring (2) over (1) is only that it's possible to also extend `forked` in similar way, which as far as I can tell would not be possible with (1).

I still think that if we change `controlled`, we should also change `forked`, but it's fine if you don't want to tackle it in this PR. We can do that separately.

Reply:
@appleby thanks! Here I will also correct `controlled`modifier, and later I will raise an issue regarding `forked`.